Replace #include guards with #pragma once

c4rlo opened the pull request:
- ~Make the names of `#include` guards consistent across the project (with a special category for `src/fdosecrets/*/*.h` based on prevailing usage). I have left out headers that clearly originate from other projects.~
- ~Also make the format of the comment after the `#endif` consistent.~
- ~Remove one case of `#pragma once`, replacing it with an `#include` guard.~

Per discussion below, this PR now replaces `#include` guards with `#pragma once`. This was done using the following script:

```python
#!/usr/bin/python

import pathlib
import re


def replace(match, path, is_found):
    is_found[0] = True
    begin_newline = match.group(1)
    body = match.group(2)
    print(f"{path}: fixing")
    num_end_newlines = 0
    for c in reversed(body):
        if c == "\n":
            num_end_newlines += 1
        else:
            break
    if num_end_newlines > 1:
        body = body[:(1-num_end_newlines)]
    return (f"{begin_newline}#pragma once\n{body}")


def process(path):
    content = path.read_text()
    if re.search(r"^#pragma once", content, flags=re.MULTILINE):
        print(f"{path}: ok")
        return

    is_found = [False]
    processed = re.sub(
        r"(^|\n)#ifndef .+?\n#define .+?\n(.*\n)#endif.+?\n?$",
        lambda m: replace(m, path, is_found),
        content,
        count=1,
        flags=re.DOTALL,
    )
    if not is_found[0]:
        print(f"{path}: missing")
    else:
        path.write_text(processed)


def relevant_files():
    yield from pathlib.Path("src").walk()
    yield from pathlib.Path("tests").walk()


for dirpath, dirnames, filenames in relevant_files():
    if str(dirpath) == "src":
        dirnames.remove("thirdparty")
    for name in filenames:
        path = dirpath / name
        if path.suffix in (".h", ".h.cmake") and str(path) != "src/streams/qtiocompressor.h":
            process(path)
```

## Testing strategy

TeamCity

## Type of change
- ✅ Refactor (significant modification to existing code)

droidmonkey commented:
Convince me why we shouldn't just switch everything over to `#pragma once`. @phoerious what are your thoughts?

c4rlo commented:
Yeah, that was my other thought. But it's not standardised. I haven't tried to look into which compilers do or don't support it.

droidmonkey commented:
Every compiler supports it on platforms we support. The only hold out was GCC prior to version 3.4 which is ancient.

phoerious commented:
I think most people would expect include guards and its what IDEs tend to auto generate. Besides that I don't really care what we use. Consistency would be nice.

droidmonkey commented:
I prefer `#pragma once` since it is (1) one line, (2) easy to understand, (3) doesn't require unique naming, (4) modern and hip :wink:

@c4rlo do you mind making that switch?

droidmonkey commented:
Bumping this

c4rlo commented:
I can have another look at this, yes. Happy to go with `#pragma once`.

c4rlo committed:
Replace #include guards with #pragma once

I used the following script to do this:

    #!/usr/bin/python

    import pathlib
    import re

    def replace(match, path, is_found):
        is_found[0] = True
        begin_newline = match.group(1)
        body = match.group(2)
        print(f"{path}: fixing")
        num_end_newlines = 0
        for c in reversed(body):
            if c == "\n":
                num_end_newlines += 1
            else:
                break
        if num_end_newlines > 1:
            body = body[:(1-num_end_newlines)]
        return (f"{begin_newline}#pragma once\n{body}")

    def process(path):
        content = path.read_text()
        if re.search(r"^#pragma once", content, flags=re.MULTILINE):
            print(f"{path}: ok")
            return

        is_found = [False]
        processed = re.sub(
            r"(^|\n)#ifndef .+?\n#define .+?\n(.*\n)#endif.+?\n?$",
            lambda m: replace(m, path, is_found),
            content,
            count=1,
            flags=re.DOTALL,
        )
        if not is_found[0]:
            print(f"{path}: missing")
        else:
            path.write_text(processed)

    def relevant_files():
        yield from pathlib.Path("src").walk()
        yield from pathlib.Path("tests").walk()

    for dirpath, dirnames, filenames in relevant_files():
        if str(dirpath) == "src":
            dirnames.remove("thirdparty")
        for name in filenames:
            path = dirpath / name
            if path.suffix in (".h", ".h.cmake") and str(path) != "src/streams/qtiocompressor.h":
                process(path)

c4rlo commented:
Updated to `#pragma once`.

c4rlo commented on src/autotype/PickcharsDialog.h:
```diff
@@ -1,50 +1,47 @@
-};
-
-#endif // KEEPASSXC_PICKCHARSDIALOG_H
+/*
+ *  Copyright (C) 2021 Team KeePassXC <team@keepassxc.org>
```
The diff with this file and a few others is that Windows-style line endings were converted to Unix-style line endings.

droidmonkey replied:
Not surprised these aren't fully normalized

droidmonkey commented:
Cool script!
